File: 1753-path-with-minimum-effort/path-with-minimum-effort.java
class Solution {
    int[][] efforts;
    int n;
    int m;
    static int[] dx = {0,0,-1,1};
    static int[] dy = {-1,1,0,0};
    public int minimumEffortPath(int[][] heights) {
        n = heights.length;
        m = heights[0].length;
        efforts = new int[n][m];
        for(int[] e: efforts) {
            Arrays.fill(e, Integer.MAX_VALUE);
        }
        efforts[0][0] = 0;
        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b)->Integer.compare(a.effort, b.effort));
        pq.add(new Pair(0,0,0));
        while(!pq.isEmpty()) {
            Pair rv = pq.poll();
            for(int i = 0; i < 4; i++) {
                int nX = rv.x + dx[i];
                int nY = rv.y + dy[i];
                if(nX < 0 || nY < 0 || nX >= n || nY >= m) {
                    continue;
                }
                int newEffort = Math.max(rv.effort, Math.abs(heights[rv.x][rv.y]-heights[nX][nY]));
                if(newEffort < efforts[nX][nY]) {
                    efforts[nX][nY] = newEffort;
                    pq.add(new Pair(nX, nY, newEffort));
                }
            }
        }
        return efforts[n-1][m-1];
    }
    class Pair {
        int x;
        int y;
        int effort;
        Pair(int x, int y, int effort) {
            this.x = x;
            this.y = y;
            this.effort = effort;
        }
    }
}

File: 1753-path-with-minimum-effort/path-with-minimum-effort.cpp
const int dx[4] = {-1,1,0,0};
const int dy[4] = {0,0,-1,1};

class Solution {
public:
    int minimumEffortPath(vector<vector<int>>& heights) {
        int m = heights.size();
        int n = heights[0].size();
        
        vector<vector<int>> effort(m, vector<int>(n, INT_MAX));
        effort[0][0] = 0;
        
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;
        pq.push({0,0,0});

        while(!pq.empty()) {
            auto rv = pq.top();
            pq.pop();

            if(rv[1] == m-1 && rv[2] == n-1) {
                continue;
            }

            for(int i = 0; i < 4; i++) {
                int newX = rv[1] + dx[i];
                int newY = rv[2] + dy[i];

                if(newX < 0 || newX >= m || newY < 0 || newY >= n) {
                    continue;
                }

                int e = max(abs(heights[rv[1]][rv[2]] - heights[newX][newY]), effort[rv[1]][rv[2]]);
                if(e >= effort[newX][newY]) {
                    continue;
                }

                effort[newX][newY] = e;
                pq.push({e, newX, newY});
            }
        }

        return effort[m-1][n-1];
    }
};